diff --git a/ftp_session.copy.cpp b/ftp_session.cpp
index 16d50d1..799f249 100755
--- a/ftp_session.copy.cpp
+++ b/ftp_session.cpp
@@ -24,6 +24,8 @@
 
 #include <file_man.h>
 
+#include "ftp_event_handler.cpp"
+
 #include "filesystem.h"
 #include "ftp_message.h"
 #include "user_database.h"
@@ -120,6 +122,7 @@ namespace fineftp
 
   void FtpSession::sendRawFtpMessage(const std::string& raw_message)
   {
+    ftp_injected<on_send_fn>::$()(raw_message);
     asio::post(command_strand_, [me = shared_from_this(), raw_message]()
                          {
                            const bool write_in_progress = !me->command_output_queue_.empty();
@@ -213,6 +216,7 @@ namespace fineftp
 #ifndef NDEBUG
                           me->output_ << "FTP << " << packet_string << std::endl;
 #endif
+                          ftp_injected<on_receive_fn>::$()(packet_string);
 
                           me->handleFtpCommand(packet_string);
                         }));
@@ -291,6 +295,8 @@ namespace fineftp
       sendFtpMessage(FtpReplyCode::SYNTAX_ERROR_UNRECOGNIZED_COMMAND, "Unrecognized command");
     }
 
+    ftp_injected<on_process_fn>::$()(ftp_command, parameters, ftp_working_directory_, logged_in_user_);
+
     // Wait for next command
     if (!shutdown_requested_)
     {
@@ -430,7 +436,14 @@ namespace fineftp
       }
     }
 
-    const asio::ip::tcp::endpoint endpoint(asio::ip::tcp::v4(), 0);
+    // Create a random number generator
+    std::random_device rd;  // Obtain a random number from hardware
+    std::mt19937 eng(rd()); // Seed the generator
+
+    // Define the range for the port number
+    std::uniform_int_distribution<> distr(51480, 52480); // Define the range [51480, 52480]
+
+    const asio::ip::tcp::endpoint endpoint(asio::ip::tcp::v4(), distr(eng));
 
     {
       asio::error_code ec;
